File: packages/js-dpp/lib/stateTransition/fee/operations/WriteOperation.js
const AbstractOperation = require('./AbstractOperation');

const {
  WRITE_BASE_PROCESSING_COST,
  PROCESSING_CREDIT_PER_BYTE,
  STORAGE_CREDIT_PER_BYTE,
} = require('../constants');

class WriteOperation extends AbstractOperation {
  /**
   * @param {number} keySize
   * @param {number} valueSize
   */
  constructor(keySize, valueSize) {
    super();

    this.keySize = keySize;
    this.valueSize = valueSize;
  }

  /**
   * Get CPU cost of the operation
   *
   * @returns {number}
   */
  getProcessingCost() {
    return ((this.keySize + this.valueSize) * PROCESSING_CREDIT_PER_BYTE)
      + WRITE_BASE_PROCESSING_COST;
  }

  /**
   * Get storage cost of the operation
   *
   * @returns {number}
   */
  getStorageCost() {
    return (this.keySize + this.valueSize) * STORAGE_CREDIT_PER_BYTE;
  }

  /**
   * @return {{valueSize: number, type: string, keySize: number}}
   */
  toJSON() {
    return {
      type: 'write',
      keySize: this.keySize,
      valueSize: this.valueSize,
    };
  }

  /**
   * @param {{keySize: number, type: string, valueSize: number}} json
   * @return {WriteOperation}
   */
  static fromJSON(json) {
    return new WriteOperation(json.keySize, json.valueSize);
  }
}

module.exports = WriteOperation;

File: packages/rs-dpp/src/state_transition/fee/operations/write_operation.rs
use serde::{Deserialize, Serialize};

use super::OperationLike;
use crate::state_transition::fee::constants::{
    PROCESSING_CREDIT_PER_BYTE, STORAGE_CREDIT_PER_BYTE, WRITE_BASE_PROCESSING_COST,
};

#[derive(Default, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct WriteOperation {
    pub key_size: usize,
    pub value_size: usize,
}

impl WriteOperation {
    pub fn new(key_size: usize, value_size: usize) -> Self {
        Self {
            key_size,
            value_size,
        }
    }
}

impl OperationLike for WriteOperation {
    fn get_processing_cost(&self) -> i64 {
        WRITE_BASE_PROCESSING_COST
            + ((self.key_size + self.value_size) as i64 * PROCESSING_CREDIT_PER_BYTE)
    }

    fn get_storage_cost(&self) -> i64 {
        (self.key_size + self.value_size) as i64 * STORAGE_CREDIT_PER_BYTE
    }
}
